classes/output/login.php

```diff
@@ -49,51 +53,84 @@ public static function output_login_options($wantsurl) {
             }
             $authplugin = get_auth_plugin($authname);
             $potentialidps = $authplugin->loginpage_idp_list($wantsurl);
-
             if (!empty($potentialidps)) {
                 foreach ($potentialidps as $idp) {
-                    $output .= \html_writer::start_tag('div', array('class' => 'idp-login container-fluid '. $authname));
-                    $name = '';
+                    $idpcontext = [
+                        'auth' => $authname,
+                        'name' => s($idp['name']),
+                        'url' => $idp['url']
+                    ];
                     if (!empty($idp['iconurl'])) {
-                        $name .= '<img src="' . s($idp['iconurl']) . '" width="24" height="24" class="mr-1"/>';
+                        $idpcontext['iconurl'] = s($idp['iconurl']);
                     }
-                    $name .= s($idp['name']);
-                    $attributes = ['class' => 'btn btn-secondary btn-block', 'title' => s($idp['name'])];
-                    $output .= \html_writer::link($idp['url'], $name, $attributes);
-
-                    $output .= \html_writer::end_tag('div');
-                    $idploginpath = $idp['url'];
+                    $idplist[] = (object) $idpcontext;
                     $count++;
+                    $idploginpath = $idp['url'];
                 }
             }
         }
+        $context->idplist = $idplist;
+
         // If only one IDP is available in authentication plugins then auto-redirect to it.
         $noredirect  = optional_param('noredirect', 0, PARAM_BOOL); // Don't redirect.
         if ($count === 1 && get_config('local_login', 'autoredirect')  && empty($noredirect)) {
             redirect($idploginpath);
         }
 
+        // If data exists in the cache, use it if it is still valid.
+        // We need to check this after calculating the IDP data to ensure its current.
+        $cache = cache::make('local_login', 'renderedlogin');
+        $cachehash = $cache->get('idphash');
+        $idphash = sha1(json_encode($idplist));
```

Comment:
Should this cache use more than just the idplist for comparing the cache? Just looking at settings.php a few things you might want to revalidate the cache when they're changed:

- headertext
- footertext
- backgroundimage
- template
- showmanual

As a side note - does this really need caching? Seems like it would complicate things. Do you actually get better performance with caching?

Reply:
If caching is necessary, could it be very clear. e.g. break it up into a couple of functions.

e.g.
```
function render_login_cached() {
    $params = self::get_login_params()
    $cachekey = self::get_cachekey($params)
    if(cache::get($cachekey) {
       // Return cached
    }
    $renderedlogin = self::render_login($params)
    cache::set($cachekey, $renderedlogin);
    return $renderedlogin;
}
```

Some benefits with this:

- Its clear what params do and don't contribute to the cache
- its very clear to read - much easier than having the cache login in the middle of a long function

Reply:
The reason this isnt the case currently is that we need to  prefigure out a bunch of info before we can decide if we are using a cache value. In this case, its most efficient to prebake half of the context object at the same time, then decide if we proceed or not. its a little more complex, but its a relatively simple function with a single call point and intent, so I dont think its too bad.